transaction/StronglyConnectedComponents.hpp

```diff
@@ -0,0 +1,119 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_STRONGLY_CONNECTED_COMPONENTS_HPP_
+#define QUICKSTEP_TRANSACTION_STRONGLY_CONNECTED_COMPONENTS_HPP_
+
+#include <stack>
+#include <vector>
+#include <unordered_map>
+
+#include "transaction/DirectedGraph.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for applying Tarjan's strongly connected components algorithm
+ *        to a directed graph. The class finds and stores information to
+ *        identify components and nodes-to-component mapping. For all nodes
+ *        v, w in a strongly connected component, there is a path from v to w,
+ *        and a path from w to v.
+ **/
+class StronglyConnectedComponents {
+ public:
+  /**
+   * @brief Constructor.
+   *
+   * @param directed_graph The directed graph which the algorithm runs.
+   **/
+  explicit StronglyConnectedComponents(const DirectedGraph &directed_graph);
+
+  /**
+   * @brief Getter for the component id of the node.
+   *
+   * @param node_id Id of the node in the graph.
+   * @return Id of the component to which the given node belongs.
+   **/
+  inline std::uint64_t getComponentId(DirectedGraph::node_id node_id) const {
```

Comment:
Me and @hakanmemisoglu discussed and concluded that size_t is fine for returning a count/size.